ulib/FStar.Fin.fst

```diff
@@ -40,44 +49,169 @@ let seqn (n: nat) (a: Type) = s: S.seq a {S.length s = n}
+  = reveal_opaque (`%is_reflexive) (is_reflexive eq)
+
+let trans_lemma (#a:Type) (eq: equivalence_relation a) (x y z:a)
+  : Lemma (requires (eq x y \/ eq y x) /\ (eq y z \/ eq z y))  
+          (ensures (x `eq` z) && (z `eq` x))   
+  = reveal_opaque (`%is_transitive) (is_transitive eq);
+    reveal_opaque (`%is_symmetric) (is_symmetric eq)
+ 
+let symm_lemma (#a:Type) (eq:equivalence_relation a) (x y:a) 
+  : Lemma (eq x y == eq y x) 
+  = reveal_opaque (`%is_symmetric) (is_symmetric eq) 
+
+(** (contains) predicate, but with custom comparison operation (a->a->bool) *) 
+[@@"opaque_to_smt"]
+let contains_eq #a (eq: equivalence_relation a) (s: S.seq a) (x:a)
+  = exists (i:under (S.length s)). eq x (S.index s i)
+
+let contains_eq_means_nonempty #a (eq:equivalence_relation a) (s: S.seq a) (x:a)
+  : Lemma (requires contains_eq eq s x) 
+          (ensures S.length s > 0) 
+          [SMTPat(contains_eq eq s x)] 
+  = reveal_opaque (`%contains_eq) (contains_eq eq) 
+
+(** a type of all elements present in a given sequence *)
+let items_of #a (eq: equivalence_relation a) (s: S.seq a) = x:a { contains_eq eq s x } 
+ 
+let tail_contains_eq #a (eq: equivalence_relation a) (s:S.seq a) 
+                        (x:a { contains_eq eq s x /\ ~(eq x (S.head s)) })
+  : Lemma (contains_eq eq (S.tail s) x) 
+  = let t = S.tail s in  
+    reveal_opaque (`%contains_eq) (contains_eq eq);  
+    eliminate exists (i: under (S.length s)). eq x (S.index s i)
+    returns exists (k: under (S.length t)). eq x (S.index t k) 
+    with _. assert (S.index s i == S.index t (i-1))
+
+(** retrieves the index of an element given prior knowledge of its presense 
+    unlike find function above, that returns option, this one guarantees
+    success, thus returning bare index instead. *)
+let rec find_eq #a (eq:equivalence_relation a) (s: S.seq a) (x:a { contains_eq eq s x })
+  : Tot (i: nat { (i < S.length s) /\ 
+                (x `eq` S.index s i) /\ 
+                (forall (j: under i). not (x `eq` S.index s j)) }) 
+        (decreases S.length s) 
+  = reveal_opaque (`%contains_eq) (contains_eq eq);  
+    if S.length s = 1 then 0
+    else if x `eq` S.index s 0 then 0 
+         else begin
+           tail_contains_eq eq s x; 
+           let ieq = find_eq eq (S.tail s) x in 
+           let aux (i: under (1 + ieq)) 
+             : Lemma (not (x `eq` S.index s i)) 
+             = if i > 0 
+               then assert (S.index (S.tail s) (i-1) == S.index s i) 
+           in Classical.forall_intro aux;
+           1 + ieq
+         end  
+  
+(** pigeonhole principle for setoids:
+    If we have a nonempty sequence (all), and a sequence (s),
+    and we know in advance that each item of (s) equals some 
+    item in (all), equals meaning (eq), not (==),
+    then we automatically know that there are at least
+    2 equivalent elements in (s).
+  
+    This procedure returns the first such pair. *)
 
+let rec pigeonhole_eq (#a:Type) (eq: equivalence_relation a) 
```

Comment:
Also, since you are already in this region of the code, would be nice to add an interface `FStar.Fin.fsti` that exports only the interesting details. But totally optional for this PR.